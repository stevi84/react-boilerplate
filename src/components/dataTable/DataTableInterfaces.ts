import { ColDef } from 'ag-grid-community';
import { Right } from '../../globals/RolesAndRights';

export enum CrudMode {
  CREATE = 'CREATE',
  READ = 'READ',
  UPDATE = 'UPDATE',
}

export interface EntityManager<EntityType> {
  create: () => void;
  read: () => void;
  update: (entity: EntityType) => void;
  delete: (entity: EntityType) => void;
  getEmpty: () => EntityType;
  editRight: Right;
}

export interface DataTableOwnProps<EntityType> {
  id: string;
  columns: ColDef<EntityType>[];
  rowsData: EntityType[];
  manager: EntityManager<EntityType>;
}
