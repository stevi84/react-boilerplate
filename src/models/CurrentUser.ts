import { Components } from '../apis/api';

export type CurrentUser = Components.Schemas.CurrentUser;
export type Role = Components.Schemas.Role;
export const getEmptyCurrentUser = (): CurrentUser => ({
  loginName: '',
  firstName: '',
  surName: '',
  mail: '',
  roles: [],
});
