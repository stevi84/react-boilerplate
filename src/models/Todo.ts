import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
