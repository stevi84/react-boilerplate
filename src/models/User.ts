import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
