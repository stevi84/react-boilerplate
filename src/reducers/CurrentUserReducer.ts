import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './Store';
import { CurrentUser, getEmptyCurrentUser } from '../models/CurrentUser';

export const currentUserInitialState: CurrentUser = getEmptyCurrentUser();

const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: currentUserInitialState,
  reducers: {
    setCurrentUser(state, action: PayloadAction<CurrentUser>) {
      return action.payload;
    },
  },
});

const { actions, reducer } = currentUserSlice;
export const { setCurrentUser } = actions;
export { reducer as currentUserReducer };

export const currentUserSelector = (state: RootState): CurrentUser => state?.currentUser || currentUserInitialState;
