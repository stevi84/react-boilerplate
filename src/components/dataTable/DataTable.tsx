import { MutableRefObject, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { DataTableOwnProps } from './DataTableInterfaces';
import ReplayIcon from '@mui/icons-material/Replay';
import AddIcon from '@mui/icons-material/Add';
import { Box, Button, Tooltip } from '@mui/material';
import { AgGridReact } from 'ag-grid-react';
import { BaseEntity } from '../../models/BaseEntity';
import 'ag-grid-community/styles/ag-grid.css';
import 'ag-grid-community/styles/ag-theme-alpine.css';
import { ActionsRenderer } from './ActionsRenderer';
import { useAppSelector } from '../../reducers/Store';
import { isReadingSelector, isSubmittingSelector } from '../../reducers/ApiCallsReducer';
import { Working } from '../common/Working';

export const DataTable = <EntityType extends BaseEntity>(props: DataTableOwnProps<EntityType>) => {
  const { id, columns, rowsData, manager } = props;

  const isReading: boolean = useAppSelector(isReadingSelector);
  const isSubmitting: boolean = useAppSelector(isSubmittingSelector);
  const { t } = useTranslation();
  const gridRef = useRef<AgGridReact>() as MutableRefObject<AgGridReact>;

  return (
    <>
      <Tooltip title={isReading ? t('tooltip_reading') : isSubmitting ? t('tooltip_submitting') : ''}>
        <span>
          <Button
            id={`${id}-button-reload`}
            variant="outlined"
            startIcon={<ReplayIcon />}
            onClick={manager.read}
            disabled={isReading || isSubmitting}
            sx={{ margin: 1 }}
          >
            {t('reload')}
          </Button>
        </span>
      </Tooltip>
      <Tooltip title={isReading ? t('tooltip_reading') : isSubmitting ? t('tooltip_submitting') : ''}>
        <span>
          <Button
            id={`${id}-button-create`}
            variant="outlined"
            startIcon={<AddIcon />}
            onClick={manager.create}
            disabled={isReading || isSubmitting}
            sx={{ margin: 1 }}
          >
            {t('create')}
          </Button>
        </span>
      </Tooltip>
      {!isReading && !isSubmitting && (
        <Box className="ag-theme-alpine" sx={{ padding: 1 }}>
          <AgGridReact<EntityType>
            ref={gridRef}
            rowData={rowsData}
            columnDefs={[
              ...columns,
              {
                field: 'actions',
                cellRenderer: ActionsRenderer,
                cellRendererParams: {
                  id: 'todo-table',
                  actionsActive: true,
                  manager,
                },
              } as any,
            ]}
            defaultColDef={{ sortable: true }}
            domLayout="autoHeight"
          />
        </Box>
      )}
      {(isReading || isSubmitting) && <Working isReading isSubmitting />}
    </>
  );
};
