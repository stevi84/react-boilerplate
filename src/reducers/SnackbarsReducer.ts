import { PrepareAction, createAction, createReducer } from '@reduxjs/toolkit';
import { OptionsObject, SnackbarKey, SnackbarMessage } from 'notistack';
import { RootState } from './Store';

// https://stackoverflow.com/questions/62754584/how-can-i-use-notistick-any-snackbar-with-redux-toolkit-and-react
// https://codesandbox.io/s/github/iamhosseindhv/notistack/tree/master/examples/redux-example?file=/redux/actions.js:0-150

interface Snackbar {
  message: SnackbarMessage;
  options?: OptionsObject;
  key: SnackbarKey;
  dismissed: boolean;
}

interface SnackbarState {
  notifications: Snackbar[];
}

export const enqueueSnackbar = createAction<PrepareAction<{ notification: Snackbar }>>(
  'ENQUEUE_SNACKBAR',
  (message: SnackbarMessage, options?: OptionsObject) => ({
    payload: {
      notification: {
        message,
        options,
        key: (options && options.key) || new Date().getTime() + Math.random(),
        dismissed: false,
      },
    },
  })
);

export const closeSnackbar = createAction<PrepareAction<{ key?: SnackbarKey; dismissAll: boolean }>>(
  'CLOSE_SNACKBAR',
  (key?: SnackbarKey) => ({
    payload: {
      dismissAll: !key, // dismiss all if no key has been defined
      key,
    },
  })
);

export const removeSnackbar = createAction<PrepareAction<{ key: SnackbarKey }>>(
  'REMOVE_SNACKBAR',
  (key: SnackbarKey) => ({
    payload: {
      key,
    },
  })
);

export const snackbarsInitialState: SnackbarState = {
  notifications: [],
};

export const snackbarsReducer = createReducer(snackbarsInitialState, (builder) =>
  builder
    .addCase(enqueueSnackbar, (state, action) => ({
      notifications: [...state.notifications, { ...action.payload.notification }],
    }))
    .addCase(closeSnackbar, (state, action) => ({
      notifications: state.notifications.map((notification) =>
        action.payload.dismissAll || notification.key === action.payload.key
          ? { ...notification, dismissed: true }
          : { ...notification }
      ),
    }))
    .addCase(removeSnackbar, (state, action) => ({
      notifications: state.notifications.filter((n) => n.key !== action.payload.key),
    }))
);

export const notificationsSelector = (state: RootState): Snackbar[] => state?.snackbar?.notifications || [];
