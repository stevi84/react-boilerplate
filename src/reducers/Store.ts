import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { todosReducer } from './TodosReducer';
import { apiCallsReducer } from './ApiCallsReducer';
import { snackbarsReducer } from './SnackbarsReducer';
import { usersReducer } from './UsersReducer';
import { currentUserReducer } from './CurrentUserReducer';

export const store = configureStore({
  reducer: {
    snackbar: snackbarsReducer,
    todos: todosReducer,
    users: usersReducer,
    apiCalls: apiCallsReducer,
    currentUser: currentUserReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
