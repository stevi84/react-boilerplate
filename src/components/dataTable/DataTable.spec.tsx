import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { EntityManager } from './DataTableInterfaces';
import { BaseEntity } from '../../models/BaseEntity';
import { DataTable } from './DataTable';
import { currentUserEditor, currentUserReader } from '../../../test/data/CurrentUser';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { todo1 } from '../../../test/data/Todo';
import { user1 } from '../../../test/data/User';
import { RootState } from '../../reducers/Store';
import { Provider } from 'react-redux';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

vi.mock('../../thunks/CurrentUserThunks', () => ({
  readCurrentUser: () => ({ type: 'readCurrentUser' }),
}));

const mockStore = configureStore([thunk]);
const initialState = {
  todos: [todo1],
  users: [user1],
  apiCalls: { runningReads: 0, runningSubmits: 0 },
  currentUser: currentUserEditor,
} as RootState;

const createMock = vi.fn();
const readMock = vi.fn();
const managerMock = {
  create: createMock,
  read: readMock,
  editRight: 'TODO_EDIT',
} as unknown as EntityManager<BaseEntity>;

describe('DataTable', () => {
  it('should equal saved snapshot', () => {
    const store = mockStore(initialState);
    const tree = render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    ).asFragment();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger read on button click', () => {
    const store = mockStore(initialState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(readMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('ReplayIcon'));
    expect(readMock.mock.calls.length).toEqual(1);
  });

  it('should trigger create on button click', () => {
    const store = mockStore(initialState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(createMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('AddIcon'));
    expect(createMock.mock.calls.length).toEqual(1);
  });

  it('should show grid if not reading or submitting', () => {
    const store = mockStore(initialState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(screen.getByRole('treegrid')).toBeTruthy();
    expect(screen.queryByRole('progressbar')).toBeFalsy();
  });

  it('should show working if reading', () => {
    const store = mockStore({ ...initialState, apiCalls: { runningReads: 1, runningSubmits: 0 } } as RootState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(screen.queryByRole('treegrid')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should show working if submitting', () => {
    const store = mockStore({ ...initialState, apiCalls: { runningReads: 0, runningSubmits: 1 } } as RootState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(screen.queryByRole('treegrid')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should hide create button if not authorized', () => {
    const store = mockStore({ ...initialState, currentUser: currentUserReader } as RootState);
    render(
      <Provider store={store}>
        <DataTable id="id" columns={[]} rowsData={[]} manager={managerMock} />
      </Provider>
    );
    expect(screen.queryByTestId('AddIcon')).toBeFalsy();
  });
});
