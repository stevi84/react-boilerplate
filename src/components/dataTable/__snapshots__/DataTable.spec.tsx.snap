// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataTable > should equal saved snapshot 1`] = `
<DocumentFragment>
  <span
    aria-label=""
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-16ww9kz-MuiButtonBase-root-MuiButton-root"
      id="id-button-reload"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ReplayIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
          />
        </svg>
      </span>
      reload
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </span>
  <span
    aria-label=""
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-16ww9kz-MuiButtonBase-root-MuiButton-root"
      id="id-button-create"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      create
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </span>
  <div
    class="ag-theme-alpine MuiBox-root css-hpgf8j"
  >
    <div
      style="height: 100%;"
    >
      <!-- AG Grid -->
      <div
        class="ag-root-wrapper ag-ltr ag-layout-auto-height"
        grid-id="1"
        role="presentation"
      >
        <div
          class="ag-root-wrapper-body ag-focus-managed ag-layout-auto-height"
          role="presentation"
        >
          <div
            class="ag-tab-guard ag-tab-guard-top"
            role="presentation"
            tabindex="0"
          />
          <!-- AG Grid Body -->
          <div
            aria-colcount="1"
            aria-rowcount="1"
            class="ag-root ag-unselectable ag-layout-auto-height"
            role="treegrid"
          >
            <div
              class="ag-header ag-pivot-off ag-header-allow-overflow"
              role="presentation"
              style="height: 49px; min-height: 49px;"
            >
              <div
                aria-hidden="true"
                class="ag-pinned-left-header ag-hidden"
                role="presentation"
                style="width: 0px; min-width: 0px; max-width: 0px;"
              />
              <div
                class="ag-header-viewport "
                role="presentation"
              >
                <div
                  class="ag-header-container"
                  role="rowgroup"
                  style="width: 200px;"
                >
                  <div
                    aria-rowindex="1"
                    class="ag-header-row ag-header-row-column"
                    role="row"
                    style="height: 48px; top: 0px; width: 200px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-description="Press ENTER to sort."
                      aria-sort="none"
                      class="ag-header-cell ag-header-cell-sortable ag-focus-managed"
                      col-id="actions"
                      role="columnheader"
                      style="width: 200px; top: 0px; height: 48px; left: 0px;"
                      tabindex="-1"
                    >
                      <div
                        aria-hidden="true"
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        aria-hidden="true"
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          aria-hidden="true"
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-4-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-header-cell-comp-wrapper"
                        role="presentation"
                      >
                        <div
                          class="ag-cell-label-container"
                          role="presentation"
                        >
                          
            
            
                          <div
                            class="ag-header-cell-label"
                            ref="eLabel"
                            role="presentation"
                          >
                            
                
                            <span
                              class="ag-header-cell-text"
                              ref="eText"
                            >
                              Actions
                            </span>
                            
                
                            <span
                              aria-hidden="true"
                              class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                              ref="eFilter"
                            >
                              <span
                                class="ag-icon ag-icon-filter"
                                role="presentation"
                                unselectable="on"
                              />
                            </span>
                            
                
                            <!--AG-SORT-INDICATOR-->
                            <span
                              class="ag-sort-indicator-container"
                              ref="eSortIndicator"
                            >
                              
            
                              <span
                                aria-hidden="true"
                                class="ag-sort-indicator-icon ag-sort-order ag-hidden"
                                ref="eSortOrder"
                              />
                              
            
                              <span
                                aria-hidden="true"
                                class="ag-sort-indicator-icon ag-sort-ascending-icon ag-hidden"
                                ref="eSortAsc"
                              >
                                <span
                                  class="ag-icon ag-icon-asc"
                                  role="presentation"
                                  unselectable="on"
                                />
                              </span>
                              
            
                              <span
                                aria-hidden="true"
                                class="ag-sort-indicator-icon ag-sort-descending-icon ag-hidden"
                                ref="eSortDesc"
                              >
                                <span
                                  class="ag-icon ag-icon-desc"
                                  role="presentation"
                                  unselectable="on"
                                />
                              </span>
                              
            
                              <span
                                aria-hidden="true"
                                class="ag-sort-indicator-icon ag-sort-mixed-icon ag-hidden"
                                ref="eSortMixed"
                              >
                                <span
                                  class="ag-icon ag-icon-none"
                                  role="presentation"
                                  unselectable="on"
                                />
                              </span>
                              
            
                              <span
                                aria-hidden="true"
                                class="ag-sort-indicator-icon ag-sort-none-icon ag-hidden"
                                ref="eSortNone"
                              >
                                <span
                                  class="ag-icon ag-icon-none"
                                  role="presentation"
                                  unselectable="on"
                                />
                              </span>
                              
        
                            </span>
                            
            
                          </div>
                          
        
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="ag-pinned-right-header ag-hidden"
                role="presentation"
                style="width: 0px; min-width: 0px; max-width: 0px;"
              />
            </div>
            <!-- AG Pinned Top -->
            <div
              class="ag-floating-top"
              role="presentation"
              style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
            >
              <!-- AG Row Container topLeft -->
              <div
                aria-hidden="true"
                class="ag-pinned-left-floating-top ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container topCenter -->
              <div
                class="ag-floating-top-viewport"
                role="presentation"
              >
                <div
                  class="ag-floating-top-container"
                  role="presentation"
                  style="width: 200px;"
                />
              </div>
              <!-- AG Row Container topRight -->
              <div
                aria-hidden="true"
                class="ag-pinned-right-floating-top ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container topFullWidth -->
              <div
                class="ag-floating-top-full-width-container"
                role="presentation"
              />
            </div>
            <div
              class="ag-body ag-layout-auto-height"
              role="presentation"
            >
              <!-- AG Middle -->
              <div
                class="ag-body-viewport ag-row-no-animation ag-layout-auto-height"
                role="presentation"
              >
                <!-- AG Row Container left -->
                <div
                  aria-hidden="true"
                  class="ag-pinned-left-cols-container ag-hidden"
                  role="presentation"
                  style="height: 1px; width: 0px; max-width: 0px; min-width: 0px;"
                />
                <!-- AG Row Container center -->
                <div
                  class="ag-center-cols-viewport"
                  role="presentation"
                  style="height: 1px;"
                >
                  <div
                    class="ag-center-cols-container"
                    role="presentation"
                    style="width: 200px; height: 1px;"
                  />
                </div>
                <!-- AG Row Container right -->
                <div
                  aria-hidden="true"
                  class="ag-pinned-right-cols-container ag-hidden"
                  role="presentation"
                  style="height: 1px; width: 0px; max-width: 0px; min-width: 0px;"
                />
                <!-- AG Row Container fullWidth -->
                <div
                  class="ag-full-width-container"
                  role="presentation"
                  style="height: 1px;"
                />
              </div>
              <!-- AG Fake Vertical Scroll -->
              <div
                aria-hidden="true"
                class="ag-body-vertical-scroll"
              >
                
            
                <div
                  class="ag-body-vertical-scroll-viewport"
                  ref="eViewport"
                >
                  
                
                  <div
                    class="ag-body-vertical-scroll-container"
                    ref="eContainer"
                    style="height: 1px;"
                  />
                  
            
                </div>
                
        
              </div>
            </div>
            <!-- AG Sticky Top -->
            <div
              class="ag-sticky-top"
              role="presentation"
              style="height: 0px; top: 49px; width: 100%;"
            >
              <!-- AG Row Container stickyTopLeft -->
              <div
                aria-hidden="true"
                class="ag-pinned-left-sticky-top ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container stickyTopCenter -->
              <div
                class="ag-sticky-top-viewport"
                role="presentation"
              >
                <div
                  class="ag-sticky-top-container"
                  role="presentation"
                  style="width: 200px;"
                />
              </div>
              <!-- AG Row Container stickyTopRight -->
              <div
                aria-hidden="true"
                class="ag-pinned-right-sticky-top ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container stickyTopFullWidth -->
              <div
                class="ag-sticky-top-full-width-container"
                role="presentation"
              />
            </div>
            <!-- AG Pinned Bottom -->
            <div
              class="ag-floating-bottom"
              role="presentation"
              style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
            >
              <!-- AG Row Container bottomLeft -->
              <div
                aria-hidden="true"
                class="ag-pinned-left-floating-bottom ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container bottomCenter -->
              <div
                class="ag-floating-bottom-viewport"
                role="presentation"
              >
                <div
                  class="ag-floating-bottom-container"
                  role="presentation"
                  style="width: 200px;"
                />
              </div>
              <!-- AG Row Container bottomRight -->
              <div
                aria-hidden="true"
                class="ag-pinned-right-floating-bottom ag-hidden"
                role="presentation"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container bottomFullWidth -->
              <div
                class="ag-floating-bottom-full-width-container"
                role="presentation"
              />
            </div>
            <!-- AG Fake Horizontal Scroll -->
            <div
              aria-hidden="true"
              class="ag-body-horizontal-scroll"
            >
              
            
              <div
                class="ag-horizontal-left-spacer"
                ref="eLeftSpacer"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              
            
              <div
                class="ag-body-horizontal-scroll-viewport"
                ref="eViewport"
              >
                
                
                <div
                  class="ag-body-horizontal-scroll-container"
                  ref="eContainer"
                  style="width: 200px;"
                />
                
            
              </div>
              
            
              <div
                class="ag-horizontal-right-spacer"
                ref="eRightSpacer"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              
        
            </div>
            <!-- AG Overlay Wrapper -->
            <div
              aria-hidden="true"
              class="ag-overlay"
            >
              
            
              <div
                class="ag-overlay-panel"
              >
                
                
                <div
                  class="ag-overlay-wrapper ag-layout-auto-height ag-overlay-no-rows-wrapper"
                  ref="eOverlayWrapper"
                >
                  <span
                    class="ag-overlay-no-rows-center"
                  >
                    No Rows To Show
                  </span>
                </div>
                
            
              </div>
              
        
            </div>
          </div>
          <div
            class="ag-tab-guard ag-tab-guard-bottom"
            role="presentation"
            tabindex="0"
          />
        </div>
        <div
          aria-hidden="true"
          class="ag-paging-panel ag-unselectable ag-hidden"
          id="ag-5"
        >
          
                
          <span
            class="ag-paging-row-summary-panel"
            role="status"
          >
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-5-first-row"
              ref="lbFirstRowOnPage"
            />
            
                    
            <span
              id="ag-5-to"
            >
              to
            </span>
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-5-last-row"
              ref="lbLastRowOnPage"
            />
            
                    
            <span
              id="ag-5-of"
            >
              of
            </span>
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-5-row-count"
              ref="lbRecordCount"
            />
            
                
          </span>
          
                
          <span
            class="ag-paging-page-summary-panel"
            role="presentation"
          >
            
                    
            <div
              aria-label="First Page"
              class="ag-button ag-paging-button"
              ref="btFirst"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-first"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <div
              aria-label="Previous Page"
              class="ag-button ag-paging-button"
              ref="btPrevious"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-previous"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <span
              class="ag-paging-description"
              role="status"
            >
              
                        
              <span
                id="ag-5-start-page"
              >
                Page
              </span>
              
                        
              <span
                class="ag-paging-number"
                id="ag-5-start-page-number"
                ref="lbCurrent"
              />
              
                        
              <span
                id="ag-5-of-page"
              >
                of
              </span>
              
                        
              <span
                class="ag-paging-number"
                id="ag-5-of-page-number"
                ref="lbTotal"
              />
              
                    
            </span>
            
                    
            <div
              aria-label="Next Page"
              class="ag-button ag-paging-button"
              ref="btNext"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-next"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <div
              aria-label="Last Page"
              class="ag-button ag-paging-button"
              ref="btLast"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-last"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                
          </span>
          
            
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
